### ISSUE TITLE
Semisynchronous replication

### ISSUE THREAD

**Issue by rtsisyk** [tarantool/tarantool#2796]
Please read https://dev.mysql.com/doc/refman/5.7/en/replication-semisync.html carefully:

<blockquote>
 Semisynchronous replication can be used as an alternative to asynchronous replication:

- A slave indicates whether it is semisynchronous-capable when it connects to the master.

- If semisynchronous replication is enabled on the master side and there is at least one semisynchronous slave, a thread that performs a transaction commit on the master blocks and waits until at least one semisynchronous slave acknowledges that it has received all events for the transaction, or until a timeout occurs.

- The slave acknowledges receipt of a transaction's events only after the events have been written to its relay log and flushed to disk.

- If a timeout occurs without any slave having acknowledged the transaction, the master reverts to asynchronous replication. When at least one semisynchronous slave catches up, the master returns to semisynchronous replication.

- Semisynchronous replication must be enabled on both the master and slave sides. If semisynchronous replication is disabled on the master, or enabled on the master but on no slaves, the master uses asynchronous replication. 

While the master is blocking (waiting for acknowledgment from a slave), it does not return to the session that performed the transaction. When the block ends, the master returns to the session, which then can proceed to execute other statements. At this point, the transaction has committed on the master side, and receipt of its events has been acknowledged by at least one slave. 
</blockquote>

There are two operational modes:

<blockquote>
 AFTER_SYNC (the default): The master writes each transaction to its binary log and the slave, and syncs the binary log to disk. The master waits for slave acknowledgment of transaction receipt after the sync. Upon receiving acknowledgment, the master commits the transaction to the storage engine and returns a result to the client, which then can proceed.

 **AFTER_COMMIT**: The master writes each transaction to its binary log and the slave, syncs the binary log, and commits the transaction to the storage engine. The master waits for slave acknowledgment of transaction receipt after the commit. Upon receiving acknowledgment, the master returns a result to the client, which then can proceed. 
</blockquote>

I propose to implement semisync replication in **AFTER_COMMIT** mode. The key point here is that master **doesn't** rollback transaction in case of error on replica (timeout, I/O, etc.). Therefore, this feature doesn't require any changes in our WAL format and can be implemented as a part of 1.7.x. Since we already have acknowledges from the replica, this feature is a piece of cake. ** AFTER_SYNC** can be added later after extending WAL format for two-phase commit.

**Comment by rtsisyk** [comment]
Need feedback from @racktear, @kostja, @GeorgyKirichenko

**Comment by rtsisyk** [comment]
AFTER SYNC can also be implemented without WAL changes using cascade rollback + WAL truncate, but I don't like this approach.

**Comment by racktear** [comment]
We need in addition:

* a way to control sync timeout
* a way to differentiate when insert/update/delete operations fail due to timeout or terminated connection to the slave
* a way to make transactions arrive to slaves in one batch, to prevent the case of partial updates

**Comment by rtsisyk** [comment]
> a way to control sync timeout

There is box.cfg { replication_time = <sec> } option.

> a way to differentiate when insert/update/delete operations fail due to timeout or terminated connection to the slave

AFAIK, master already sets different messages in box.info.replication[x].message for "connection reset by peer"  and  timeouts cases. I planned to add `.error: object` instead of `.message: string` to allow matching by error code (error.type == "ClientError" or error.type == "Timeout") instead of by error messages.

> a way to make transactions arrive to slaves in one batch, to prevent the case of partial updates

Definitely makes sense. It is a serious gotcha in 1.6+ replication protocol which also blocks parallel applier. I filed a separate ticket for this issue. https://github.com/tarantool/tarantool/issues/2798

**Comment by sergos** [comment]
Quorum sync replication covers both AFTER_SYNC and AFTER_COMMIT with data visibility in engine controlled via MVCC option. Consider this as done.
